src/codecs/framing/character_delimited.rs

```diff
@@ -87,80 +81,51 @@ impl Decoder for CharacterDelimitedDecoder {
+                        // The discovered sub-buffer is too big, so we discard
+                        // it, taking care to also discard the delimiter.
+                        warn!(
+                            message = "Discarding frame larger than max_length.",
+                            buf_len = buf.len(),
+                            max_length = self.max_length,
+                            internal_log_rate_secs = 30
+                        );
+                        buf.advance(next_delimiter_idx + 1);
+                    } else {
+                        let frame = buf.split_to(next_delimiter_idx).freeze();
+                        trace!(
+                            message = "Decoding the frame.",
+                            bytes_proccesed = frame.len()
+                        );
+                        buf.advance(1); // scoot past the delimiter
+                        return Ok(Some(frame));
                     }
                 }
-                (false, Some(pos)) => {
-                    // We found a correct frame
-
-                    let newpos_index = pos + self.next_index;
-                    self.next_index = 0;
-                    let mut frame = buf.split_to(newpos_index + 1);
-
-                    trace!(
-                        message = "Decoding the frame.",
-                        bytes_proccesed = frame.len()
-                    );
-
-                    let frame = frame.split_to(frame.len() - 1);
-
-                    return Ok(Some(frame.freeze()));
-                }
-                (false, None) if buf.len() > self.max_length => {
-                    // We reached the max length without finding the
-                    // delimiter so must discard the rest until we
-                    // reach the next delimiter
-                    self.is_discarding = true;
-                    warn!(
-                        message = "Discarding frame larger than max_length.",
-                        buf_len = buf.len(),
-                        max_length = self.max_length,
-                        internal_log_rate_secs = 30
-                    );
-                }
-                (false, None) => {
-                    // We didn't find the delimiter and didn't
-                    // reach the max frame length.
-                    self.next_index = read_to;
-                    return Ok(None);
-                }
             }
         }
     }
 
     fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Bytes>, Self::Error> {
-        let frame = self.decode(buf)?.or_else(|| {
-            (!buf.is_empty() && !self.is_discarding).then(|| {
-                self.next_index = 0;
-                buf.split_to(buf.len()).into()
-            })
-        });
-
-        Ok(frame)
+        match self.decode(buf)? {
+            Some(frame) => Ok(Some(frame)),
+            None => {
+                if buf.is_empty() || buf.len() > self.max_length {
```

Comment:
👍 One note: I think we should have the same warning we have above about discarding frames for the `buf.len() > self.max_length` case.

Reply:
Addressed in 90e6cd4fb